middleware/ack.go

```diff
@@ -3,15 +3,46 @@ package middleware
 import (
 	"context"
 
-	amqp "github.com/rabbitmq/amqp091-go"
-
 	amqprpc "github.com/0x4b53/amqp-rpc/v3"
+	amqp "github.com/rabbitmq/amqp091-go"
 )
 
+// OnErrFunc is the function that will be called when the middleware get an
+// error from `Ack`. The error and the delivery will be passed.
+type OnErrFunc func(err error, delivery amqp.Delivery)
+
+// AckLogError is a built-in function that will log the error if any is returned
+// from `Ack`.
+//
+//	middleware := AckDelivery(AckLogError(log.Printf))
+func AckLogError(logFn amqprpc.LogFunc) OnErrFunc {
+	return func(err error, delivery amqp.Delivery) {
+		logFn("could not ack delivery (%s): %v\n", delivery.CorrelationId, err)
```

Comment:
Nice catch! I opted to use `\n` since `LogFunc` signature is `type LogFunc func(format string, args ...interface{})` which is implemented by e.g. `fmt.Printf`, `log.Printf` which does not add a trailing newline. Methods like `logrus.Infof` will add a newline but if it's included in the message it will be printed as `\n` and not an actual newline.

[Example](https://go.dev/play/p/zj0rE-ooX2v):

```go
func main() {
	fmt.Printf("Before")

	l := logrus.New()
	l.Infof("hello %s\n", "world")

	fmt.Printf("After")
}
```

```sh
Beforetime="2009-11-10T23:00:00Z" level=info msg="hello world\n"
After
```

Do you think this makes sense or should I remove it?

Reply:
Makes sense 👍